crates/sui-framework/packages/dwallet-system/sources/dwallet.move

```diff
@@ -310,4 +311,77 @@ module dwallet_system::dwallet {
         };
         transfer::transfer(sign_output, session.sender);
     }
+
+    /// Encrypt DWallet secret share with an AHE public key.
+    const EEncryptUserShare: u64 = 0x1;
+    const EInvalidEncryptionKeyScheme: u64 = 0x2;
+
+    struct EncryptedUserShare has key {
+        id: UID,
+        dwallet_id: ID,
+        encrypted_secret_share_and_proof: vector<u8>,
+        encryption_key_id: ID,
+    }
+
+    /// An Additively Homomorphic Encryption (AHE) public key
+    /// that can be used to encrypt a user share in order to prove to the network that
+    /// the recipient can sign with a dWallet when it is transferred or access is granted to it.
+    struct EncryptionKey has key {
+        id: UID,
+        scheme: u8,
+        encryption_key: vector<u8>,
+        key_owner_address: address,
+    }
+
+    const PaillierEncryptionKey: u8 = 0;
+    fun is_valid_encryption_key_scheme(scheme: u8): bool {
+        scheme == PaillierEncryptionKey // || scheme == ...
+    }
+
+    public fun register_encryption_key(key: vector<u8>, scheme: u8, ctx: &mut TxContext): ID {
+        assert!(is_valid_encryption_key_scheme(scheme), EInvalidEncryptionKeyScheme);
+        let encryption_key = EncryptionKey {
+            id: object::new(ctx),
+            scheme: scheme,
+            encryption_key: key,
+            key_owner_address: tx_context::sender(ctx),
+        };
+        let encryption_key_id = object::id(&encryption_key);
+        transfer::freeze_object(encryption_key);
+        encryption_key_id
+    }
+
+    public fun encrypt_user_share(
+        dwallet: &DWallet,
+        encryption_key: &EncryptionKey,
+        encrypted_secret_share_and_proof: vector<u8>,
+        recipient: address,
+        ctx: &mut TxContext,
+    ): ID {
+        let is_valid = validate_encrypted_user_secret_share(
+            encryption_key.encryption_key,
+            encrypted_secret_share_and_proof,
+            output(dwallet),
+        );
+
+        assert!(is_valid, EEncryptUserShare);
+
+        let encrypt_user_share = EncryptedUserShare {
+            id: object::new(ctx),
+            dwallet_id: object::id(dwallet),
+            encrypted_secret_share_and_proof,
+            encryption_key_id: object::id(encryption_key),
+        };
+
+        let encrypt_user_share_obj_id = object::id(&encrypt_user_share);
+        transfer::transfer(encrypt_user_share, recipient);
```

Comment:
I prefer having it as an immutable object, but in any case if it is owned you must check that `address == encryption_key.key_onwer_address`